docs: update edge functions integration guides with new server package (#45058)
supabase/supabase: apps/docs/content/guides/functions/auth.mdx

---
id: 'auth'
title: 'Securing Edge Functions'
description: 'Authentication patterns for Supabase Edge Functions.'
subtitle: 'Authentication patterns for Edge Functions'
---

<Admonition type="caution">

> The patterns in this guide assume your project uses the new [JWT signing keys](/docs/guides/auth/signing-keys) and the new [API keys](https://github.com/orgs/supabase/discussions/29260). If you're still on legacy JWTs, see the [Legacy JWT Secret guide](/docs/guides/functions/auth-legacy-jwt).

Reviewer: ```suggestion
The patterns in this guide assume your project uses the new [JWT signing keys](/docs/guides/getting-started/api-keys) and the new [API keys](https://github.com/orgs/supabase/discussions/29260). If you're still on legacy JWTs, see the [Legacy JWT Secret guide](/docs/guides/functions/auth-legacy-jwt).
```

I have some changes in process to this page that I think would make it a much better place to direct people to. It will be moved to the path above.

</Admonition>

Every request to an Edge Function passes through two layers of auth. First, a platform-level check (`verify_jwt`) runs before your code executes. Then, once the request reaches your handler, you decide what to do with the credentials the caller sent.

## Understanding authorization headers

> Edge Functions care about two request headers. Sending the wrong credential in the wrong header is the most common source of 401s.

Reviewer: ```suggestion
Edge Functions care about two request headers. Sending the wrong credential in the wrong header is the most common source of 401 errors.
```

| Header          | Value                                   | Used for                               |
| --------------- | --------------------------------------- | -------------------------------------- |
| `Authorization` | `Bearer <user-jwt>`                     | A user signed in through Supabase Auth |
| `apikey`        | `sb_publishable_...` or `sb_secret_...` | Calls from clients or services         |

> A common mistake is sending a publishable or secret key as a bearer token: `Authorization: Bearer sb_publishable_...`. The new API keys are not JWTs. The platform check can't validate them, and your handler can't verify them as JWTs either. Put API keys in the `apikey` header.

Reviewer: ```suggestion
A common mistake is sending a publishable or secret key as a bearer token: `Authorization: Bearer sb_publishable_...`. The new API keys are not JWTs. The platform check can't validate them, and your handler can't verify them as JWTs either. Instead, put API keys in the `apikey` header.
```

> Both headers can travel together. A signed-in user calling your function through `supabase-js`, for example, sends their session JWT in `Authorization` and the project's publishable key in `apikey`.

Reviewer: ```suggestion
You can send both headers together. A signed-in user calling your function through `supabase-js`, for example, sends their session JWT in `Authorization` and the project's publishable key in `apikey`.
```

## The `verify_jwt` platform check

> When `verify_jwt` is on (the default), the platform inspects the `Authorization` header of every request before your function runs. It expects a valid user JWT. If the header is missing, malformed, or signed with a different key, the platform returns a 401 and your code never executes.

Reviewer: ```suggestion
When `verify_jwt` is enabled (the default), the platform inspects the `Authorization` header of every request before your function runs. It expects a valid user JWT. If the header is missing, malformed, or signed with a different key, the platform returns a 401 error, and your code never executes.
```

> The check validates legacy HS256 JWTs and JWTs signed with the new asymmetric [signing keys](/docs/guides/auth/signing-keys). What it will not accept is an API key. Publishable and secret keys are not JWTs, so callers that send one on `Authorization` fail the check before their request reaches your handler.

Reviewer: ```suggestion
The check validates legacy HS256 JWTs and JWTs signed with the new asymmetric [signing keys](/docs/guides/auth/signing-keys).

The check does not accept an API key. Publishable and secret keys are not JWTs, so callers that send one in the `Authorization` header fail the check before their request reaches your handler.
```

> Use the flag to match how the function is called:

Reviewer: ```suggestion
Use the `verify_jwt` flag to match how the function is called:
```

- **Leave `verify_jwt` on** for functions that are only called with a user JWT, such as functions invoked from the client through `supabase.functions.invoke`. The platform rejects anonymous requests before they reach your code, and your handler can trust that a valid JWT is present.
- **Turn `verify_jwt` off** for functions that are called without an `Authorization` header, such as webhooks from external providers, or service-to-service calls that authenticate with an API key. These patterns are covered later in the guide.

> Set the flag per function in `supabase/config.toml`:

Reviewer: I think it would be a better flow to move this "how to do it" up to line 35

```toml
[functions.stripe-webhook]
verify_jwt = false
```

For 401 failure modes and how to diagnose them, see [Edge Function 401 error response](/docs/troubleshooting/edge-function-401-error-response).

## Common auth patterns

> The sections below show the four patterns you'll reach for most often, written without an SDK so the auth moves are visible. Business logic is left as a placeholder. The next section shows the same four patterns using [`@supabase/server`](https://github.com/supabase/server).

Reviewer: > written without an SDK so the auth moves are visible.

What do you mean? PSeudocode?

### Authenticated user calls

> Keep `verify_jwt` on. The platform validates the JWT before your handler runs. Forward the `Authorization` header into the Supabase client so queries run under the caller's RLS policies.

Reviewer: ```suggestion
Keep `verify_jwt` enabled. The platform validates the JWT before your handler runs. Forward the `Authorization` header to the Supabase client so queries run under the caller's RLS policies.
```

```toml
[functions.notes]
verify_jwt = true
```

```ts
import { createClient } from 'npm:@supabase/supabase-js@2'

Deno.serve((req) => {
  const supabase = createClient(
    Deno.env.get('SUPABASE_URL')!,

>     Deno.env.get('SUPABASE_PUBLISHABLE_KEY')!,

Reviewer 2: We don't have "default" publishable key exposed.

Instead it needs to consume from `SUPABASE_PUBLISHABLE_KEYS` — plural form.

Then it has to be Json parsed before calling `["default"]` entry

Reviewer 2: Applied at https://github.com/supabase/supabase/pull/45058/commits/c7262d4c6840933512d06bf25c2a1158de9a465c

    { global: { headers: { Authorization: req.headers.get('Authorization')! } } }
  )

  // your business logic. queries run as the caller
  return Response.json({ ok: true })
})
```

### Service-to-service calls

> Cron jobs, workers, `pg_net`, or another Edge Function call in with a secret key on the `apikey` header. These callers don't send a user JWT, so turn `verify_jwt` off and validate the key yourself.

Reviewer: ```suggestion
Cron jobs, workers, `pg_net`, or another Edge Functions make calls with a secret key on the `apikey` header. These callers don't send a user JWT, so disable `verify_jwt` and validate the key yourself.
```

```toml
[functions.run-automations]
verify_jwt = false
```

```ts
import { createClient } from 'npm:@supabase/supabase-js@2'

Deno.serve((req) => {
  if (req.headers.get('apikey') !== Deno.env.get('INTERNAL_AUTOMATIONS_KEY')) {
    return Response.json({ error: 'forbidden' }, { status: 401 })
  }

  const supabase = createClient(Deno.env.get('SUPABASE_URL')!, Deno.env.get('SUPABASE_SECRET_KEY')!)

  // your business logic. queries run with the service role
  return Response.json({ ok: true })
})
```

> Never expose a secret key to the browser. Store it as a [function secret](/docs/guides/functions/secrets).

Reviewer: Assuming this is in relation to the `INTERNAL_AUTOMATIONS_KEY` above, I think it needs to stand out way more, so maybe:

```suggestion
<Admonition type="tip">

Never expose a secret key to the browser. Store it as a [function secret](/docs/guides/functions/secrets).

<Admonition>
```

### Public functions

> For a genuinely public function, like a health check, no credential is required. Turn `verify_jwt` off so anonymous callers can reach the handler.

Reviewer: ```suggestion
For a genuinely public function, like a health check, no credential is required. Disable `verify_jwt` so anonymous callers can reach the handler.
```

```toml
[functions.health]
verify_jwt = false
```

```ts
Deno.serve(() => {
  // your business logic
  return Response.json({ ok: true })
})
```

### External webhooks

> External providers like Stripe or GitHub don't send Supabase credentials. They sign the request body with their own shared secret. Turn `verify_jwt` off and verify the signature before acting on the payload.

Reviewer: ```suggestion
External providers like Stripe or GitHub don't send Supabase credentials. They sign the request body with their own shared secret. Disable `verify_jwt` and verify the signature before acting on the payload.
```

[functions.stripe-webhook]
verify_jwt = false
```

```ts
import Stripe from 'npm:stripe'

const stripe = new Stripe(Deno.env.get('STRIPE_SECRET_KEY')!)

Deno.serve(async (req) => {
  const signature = req.headers.get('stripe-signature') ?? ''
  const body = await req.text()

  try {
    stripe.webhooks.constructEvent(body, signature, Deno.env.get('STRIPE_WEBHOOK_SECRET')!)
  } catch {
    return new Response('bad signature', { status: 400 })
  }

  // your business logic. handle the event
  return Response.json({ received: true })
})
```

## Simplifying with `@supabase/server`

The [`@supabase/server`](https://github.com/supabase/server) package wraps your handler, checks the caller's credentials against a declared `allow` mode, and hands you a pre-configured Supabase client on `ctx`. The same patterns above, written against the SDK, look like this.

| Mode              | Accepts                                    |
| ----------------- | ------------------------------------------ |
| `'user'`          | A valid user JWT on `Authorization`        |
| `'secret:<name>'` | A named secret key on `apikey`             |
| `'always'`        | Any caller, no check (for signed webhooks) |

> See the [`@supabase/server` docs](https://github.com/supabase/server) for the full list of modes.

Reviewer: ```suggestion
<Admonition type="tip">

See the [`@supabase/server` docs](https://github.com/supabase/server) for the full list of modes.

</Admonition>
```

The intention is for these to find their way into the docs?

### Authenticated user calls

`allow: 'user'` pairs with `verify_jwt = true`. The platform validates the JWT, and the SDK hands you `ctx.supabase` already scoped to the caller.

```ts
import { withSupabase } from 'npm:@supabase/server'

export default {
  fetch: withSupabase({ allow: 'user' }, async (_req, ctx) => {
    // your business logic. ctx.supabase is scoped to the caller
    return Response.json({ email: ctx.userClaims?.email })
  }),
}
```

### Service-to-service calls

`allow: 'secret:<name>'` validates the `apikey` header against the named secret key from your [dashboard](/dashboard/project/_/settings/api-keys) and gives you `ctx.supabaseAdmin` for privileged work. The `<name>` matches the name you gave the key. Keep `verify_jwt = false`.

```ts
import { withSupabase } from 'npm:@supabase/server'

export default {
  fetch: withSupabase({ allow: 'secret:automations' }, async (_req, ctx) => {
    // your business logic. ctx.supabaseAdmin bypasses RLS
    return Response.json({ ok: true })
  }),
}
```

<Admonition type="tip">

> Create a named secret key for each caller in your project's [API keys settings](/dashboard/project/_/settings/api-keys). Give it a name like `automations`, and share the generated `sb_secret_...` value with the service that calls this function.

Reviewer: ```suggestion
Create a named secret key for each caller in the [**Settings > API keys**](/dashboard/project/_/settings/api-keys) section of the Dashboard. Give it a name like "automations", and share the generated `sb_secret_...` value with the service that calls this function.
```

This is a pattern we are slowly starting to roll out.

---
id: 'auth'
title: 'Securing Edge Functions'
description: 'Authentication patterns for Supabase Edge Functions.'
subtitle: 'Authentication patterns for Edge Functions'
---

<Admonition type="warning">

The patterns in this guide assume your project uses the new [JWT signing keys](/docs/guides/auth/signing-keys) and the new [API keys](https://github.com/orgs/supabase/discussions/29260). If you're still on legacy JWTs, see the [Legacy JWT Secret guide](/docs/guides/functions/auth-legacy-jwt).

</Admonition>

Every request to an Edge Function passes through two layers of auth. First, a platform-level check (`verify_jwt`) runs before your code executes. Then, once the request reaches your handler, you decide what to do with the credentials the caller sent.

## Understanding authorization headers

> Edge Functions care about two request headers. Sending the right credential in the right header is the most common source of 401s.

Reviewer 2: > Sending the [right] credential in the [right] header is the most common source of 401s

I think it got miss written

A common mistake is sending a publishable or secret key as a bearer token: `Authorization: Bearer sb_publishable_...`. The new API keys are not JWTs. The platform check can't validate them, and your handler can't verify them as JWTs either. Put API keys in the `apikey` header.

Both headers can travel together. A signed-in user calling your function through `supabase-js`, for example, sends their session JWT in `Authorization` and the project's publishable key in `apikey`.

## The `verify_jwt` platform check

When `verify_jwt` is on (the default), the platform inspects the `Authorization` header of every request before your function runs. It expects a valid user JWT. If the header is missing, malformed, or signed with a different key, the platform returns a 401 and your code never executes.

The check validates legacy HS256 JWTs and JWTs signed with the new asymmetric [signing keys](/docs/guides/auth/signing-keys). What it will not accept is an API key. Publishable and secret keys are not JWTs, so callers that send one on `Authorization` fail the check before their request reaches your handler.

Use the flag to match how the function is called:

> - **Leave `verify_jwt` on** for functions that are only called with a user JWT, such as functions invoked from the client through `supabase.functions.invoke`. The platform rejects anonymous requests before they reach your code, and your handler can trust that a valid JWT is present.

Reviewer 3: > rejects anonymous requests

I think this is a bit ambiguous because we have a concept of anon keys

Reviewer: At the same time, we're moving away from them, but we could maybe say:

```suggestion
- **Leave `verify_jwt` on** for functions that are only called with a user JWT, such as functions invoked from the client through `supabase.functions.invoke`. The platform rejects unauthenticated requests before they reach your code, and your handler can trust that a valid JWT is present.
```

```ts
Deno.serve(() => {
  // your business logic
  return Response.json({ ok: true })
})
```

### External webhooks

External providers like Stripe or GitHub don't send Supabase credentials. They sign the request body with their own shared secret. Turn `verify_jwt` off and verify the signature before acting on the payload.

```toml
[functions.stripe-webhook]
verify_jwt = false
```

```ts
import Stripe from 'npm:stripe'

const stripe = new Stripe(Deno.env.get('STRIPE_SECRET_KEY')!)

Deno.serve(async (req) => {
  const signature = req.headers.get('stripe-signature') ?? ''
  const body = await req.text()

  try {
    stripe.webhooks.constructEvent(body, signature, Deno.env.get('STRIPE_WEBHOOK_SECRET')!)
  } catch {
    return new Response('bad signature', { status: 400 })
  }

  // your business logic. handle the event
  return Response.json({ received: true })
})
```

## Simplifying with `@supabase/server`

The [`@supabase/server`](https://github.com/supabase/server) package wraps your handler, checks the caller's credentials against a declared `allow` mode, and hands you a pre-configured Supabase client on `ctx`. The same patterns above, written against the SDK, look like this.

> | `allow` mode      | Accepts                                    |

Reviewer 2: This table is broken, `allow` is in the same line as `mode` header

Reviewer 2: Ahh I see, the intention is to have the header name as "Allow mode", but I don't think if using code span in the header is a good pattern

## Understanding authorization headers

> Edge Functions care about two request headers. Sending the wrong credential in the wrong header is the most common source of 401s.x

Reviewer 3: one more (new) typo here

Reviewer 2: Do you mean the `x` at end of line? yh seems a typo
